### PR TITLE
Update rotation logic to generate new password when a retried password fails

### PR DESCRIPTION
**Pull request by vinay-gopalan**
# Overview
Previously, when a static role rotation failed, we would store the password in the WAL entry. Upon retries, this same stored password would be reused. This leads customers to infinite loop cases where a failing password is infinitely retried.

This is a workaround for cases where updating the password policy will not help.

# Contributor Checklist
- [x] Add relevant docs to upstream Vault repository, or sufficient reasoning why docs won’t be added yet
- [x] Add output for any tests not ran in CI to the PR description (eg, acceptance tests)
- [x] Backwards compatible

# Testing

```
=== RUN   TestRoles_NewPasswordGeneration
2024-11-20T15:40:14.309-0800 [INFO]  initializing database rotation queue
2024-11-20T15:40:14.310-0800 [INFO]  populating role rotation queue
2024-11-20T15:40:14.310-0800 [DEBUG] no WAL entries found
2024-11-20T15:40:14.310-0800 [INFO]  starting periodic ticker
2024-11-20T15:40:14.310-0800 [DEBUG] wrote WAL: role=hashicorp WAL ID=31031583-48de-c64c-28ec-0a506a06c986
2024-11-20T15:40:14.310-0800 [DEBUG] deleted WAL: WAL ID=31031583-48de-c64c-28ec-0a506a06c986
=== RUN   TestRoles_NewPasswordGeneration/rotation_failures_should_generate_new_password_on_retry
2024-11-20T15:40:14.310-0800 [DEBUG] wrote WAL: role=hashicorp WAL ID=7bcc84df-5a3b-d5d4-31a1-45acbd4af9fa
2024-11-20T15:40:14.310-0800 [WARN]  unable to rotate credentials in rotate-role: error="forced error"
2024-11-20T15:40:14.310-0800 [DEBUG] password stored in WAL failed, generating new password: role=hashicorp WAL ID=7bcc84df-5a3b-d5d4-31a1-45acbd4af9fa
2024-11-20T15:40:14.310-0800 [DEBUG] wrote WAL: role=hashicorp WAL ID=9ae4a137-2029-c4ff-c4a5-cc756bc5d528
2024-11-20T15:40:14.310-0800 [WARN]  unable to rotate credentials in rotate-role: error="forced error"
2024-11-20T15:40:14.310-0800 [DEBUG] password stored in WAL failed, generating new password: role=hashicorp WAL ID=9ae4a137-2029-c4ff-c4a5-cc756bc5d528
2024-11-20T15:40:14.310-0800 [DEBUG] wrote WAL: role=hashicorp WAL ID=6f3ba486-ebf3-bd98-9206-1c6f12990ca6
2024-11-20T15:40:14.310-0800 [DEBUG] deleted WAL: WAL ID=6f3ba486-ebf3-bd98-9206-1c6f12990ca6
--- PASS: TestRoles_NewPasswordGeneration/rotation_failures_should_generate_new_password_on_retry (0.00s)
=== RUN   TestRoles_NewPasswordGeneration/updating_password_policy_should_generate_new_password
2024-11-20T15:40:14.310-0800 [DEBUG] wrote WAL: role=hashicorp WAL ID=41ee98d6-9266-74cc-bce2-5875d2924ec1
2024-11-20T15:40:14.310-0800 [WARN]  unable to rotate credentials in rotate-role: error="forced error"
2024-11-20T15:40:14.310-0800 [DEBUG] password policy changed, generating new password: role=hashicorp WAL ID=41ee98d6-9266-74cc-bce2-5875d2924ec1
2024-11-20T15:40:14.310-0800 [DEBUG] wrote WAL: role=hashicorp WAL ID=4759d581-3a2f-9d1f-201e-e91ad908ed16
2024-11-20T15:40:14.310-0800 [WARN]  unable to rotate credentials in rotate-role: error="forced error"
2024-11-20T15:40:14.310-0800 [DEBUG] password stored in WAL failed, generating new password: role=hashicorp WAL ID=4759d581-3a2f-9d1f-201e-e91ad908ed16
2024-11-20T15:40:14.310-0800 [DEBUG] wrote WAL: role=hashicorp WAL ID=2e9f6355-9322-b5e6-f36b-bd96b8cf8ee5
2024-11-20T15:40:14.310-0800 [DEBUG] deleted WAL: WAL ID=2e9f6355-9322-b5e6-f36b-bd96b8cf8ee5
--- PASS: TestRoles_NewPasswordGeneration/updating_password_policy_should_generate_new_password (0.00s)
--- PASS: TestRoles_NewPasswordGeneration (0.00s)
PASS
```

### CODE REVIEW

File: rotation.go [hunk 1]
```diff
@@ -384,6 +386,17 @@ func (b *backend) setStaticAccountPassword(ctx context.Context, s logical.Storag
 		err = b.client.UpdateUserPassword(config.LDAP, input.Role.StaticAccount.Username, newPassword)
 	}
 	if err != nil {
+		if usedExistingCredential {
+			// A retried password has failed again. Delete WAL and try with fresh password
```
Comment (fairclothjm) [inline rotation.go:390]: nit: This comment is a little unclear to me. I have context here so I understand what it is trying to say. But I am not so sure about someone reading this that doesn't have the context.

I think if we renamed `usedExistingCredential` to `usedCredentialFromPreviousRotation` then we could omit this comment entirely. If we do that, I think the comment on line 396 should be sufficient to explain the expectations going forward.

File: path_static_roles_test.go [hunk 2]
```diff
@@ -435,6 +435,126 @@ func TestRoles(t *testing.T) {
...
+		walIDs := requireWALs(t, storage, 1)
+		wal, err := b.findStaticWAL(ctx, storage, walIDs[0])
+		if err != nil || wal == nil {
+			t.Fatal(err)
+		}
+
+		// Store password
+		initialPassword := wal.NewPassword
+
+		// Rotate role manually and fail again #1
```
Comment (fairclothjm) [inline path_static_roles_test.go:462]: Shouldn't this be `#2` and Line 450 be `#1`?

File: path_static_roles_test.go [hunk 3]
```diff
@@ -435,6 +435,126 @@ func TestRoles(t *testing.T) {
...
+		generateWALFromFailedRotation(t, b, storage, roleName)
+
+		// Ensure WAL is deleted since retrying initial password failed
+		requireWALs(t, storage, 0)
+
+		// Rotate role manually and fail again #2
+		generateWALFromFailedRotation(t, b, storage, roleName)
```
Comment (fairclothjm) [inline path_static_roles_test.go:469]: Do we need this call to `generateWALFromFailedRotation`? If I am following the test setup correctly, I think the previous 2 calls to `generateWALFromFailedRotation` should be sufficient.

**Review by fairclothjm — COMMENTED** [review]
Looking good so far! Can we also add a changelong entry?

File: rotation.go (hunk 1 above)
Comment (vinay-gopalan) [reply rotation.go]: Updated in latest, thanks!

File: path_static_roles_test.go (hunk 3 above)
Comment (vinay-gopalan) [reply path_static_roles_test.go]: Removed the extra test step! Thanks for the feedback :)

**Review by fairclothjm — APPROVED** [review]
LGTM! 🚀

**Pull request merged**